### DISCUSSION TITLE
[Question] Exclusive fullscreen mode programs

### DISCUSSION THREAD

**Discussion by vanfanel** [labwc/labwc#1518] [Q&A] [answered]
Hi,

I am targeting labwc as a low latency compositor for fullscreen programs that use SDL2 or RetroArch.
I have also read that wlroots provides direct-scanout for programs running in exclusive fullscreen mode.

SDL2 provides real fullscreen mode when creating a window with `SDL_CreateWindow()` when passing the `SDL_WINDOW_FULLSCREEN` flag. I understand that should be an exclusive fullscreen window, and I shouldn't be able to swap ("ALT-TAB") between windows once that exclusive fullscreen window is created and on screen: but I am able to do so.

Same goes for RetroArch: on labwc, I can swap bewteen real fullscreen windows.

So, does labwc support exclusive fullscreen windows? Or maybe it is an wlroots feature not yet present on this compositor?

**Comment by ahesford** [comment]
Where are you seeing that `SDL_WINDOW_FULLSCREEN` implies an inability to switch to another window? The documentation I found suggested that this flag enables a full-screen view with a video mode change, not that it should also prevent compositors from swapping to another window.

You can certainly disable Alt-Tab (and all other labwc keybinds) with a `ToggleKeybinds` action to avoid interfering with client interactions.

**Reply by vanfanel** [reply]
I haven't read that `SDL_WINDOW_FULLSCREEN` implies an inability to switch to another window, but I have read that a real fullscreen window (I think the common term is "exclusive fullscreen") does prevent switching to another window.
Since I can switch to another window when I have a  window created with `SDL_WINDOW_FULLSCREEN`, I deducted that this window isn't really a really fullscreen window.

Maybe I am wrong on the concepts, that's why I am asking humbly to be illustrated :)

**Comment by Consolatis** [comment]
Direct scanout should be enabled automatically by wlroots if there is a window that matches the resolution of the monitor, e.g. usual fullscreen. Things that could prevent that from happen are other surfaces on top of that fullscreen window, for example popups from a notification daemon and even the mouse cursor (if rendered via software). Also screen capture automatically disables direct-scanout as far as I remember. So usual fullscreen and the cursor hidden should cause direct-scanout to be enabled. 

Edit:
Any kind of transform like rotation or scale of the output likely prevents it as well.
> The documentation I found suggested that this flag enables a full-screen view with a video mode change

Does the mode change actually happen? Could be verified via wlr-randr. I guess that would indeed be the most optimal setup in terms of latency if you are fine with the quality of your actual monitor doing the scaling rather than the emulator or wlroots.

**Reply by vanfanel** [reply]
The emulators and SDL2 programs don't need to have their original video mode set: using the desktop resolution is perfectly fine (scaling is done via hardware renderers: OpenGL in SDL2, Vulkan/OpenGL in RetroArch). So there's no need for a videomode change at all: the idea is to have every program running at the native display resolution.

Now, with that in mind, I can do one of these:
-Call `SDL_CreateWindow()` with `SDL_WINDOW_FULLSCREEN`  and passing the current/native resolution.
-Call `SDL_CreateWindow()` with `SDL_WINDOW_FULLSCREEN_DESKTOP`.
In both cases, I get a fullscreen window. But my understanding is that, if I use `SDL_WINDOW_FULLSCREEN_DESKTOP`, I will get a "fake" fullscreen window, hence direct-scanout is disabled.

Am I right on this?

Or maybe wlroots uses direct-scanout with every fullscreen window, even if it's not "exclusive fullscreen" ?
Maybe I am bringing up concepts that don't apply on the Wayland world and "exclusive fullscreen" windows were an X11-only thing?

Also, is there a way to find out if direct-scanout is active for a window? Anything will do: I can connect remotely via SSH to find out.

**Reply by vanfanel** [reply]
@Consolatis Also, looking at the SDL2 docs here:
https://wiki.libsdl.org/SDL2/SDL_SetWindowFullscreen#remarks
It confirms that passing `SDL_WINDOW_FULLSCREEN_DESKTOP` a "fake" fullscreen window is created. But I am not sure how that translates to wlroots/labwc, or if it makes sense on a Wayland context.

**Reply by Consolatis** [reply, marked as answer]
As far as I remember wlroots really only looks if the geometry completely matches the output geometry and then if no other surfaces are on top enables direct-scanout. As said, even a mouse cursor when rendered by wlroots can prevent that from happen, depending on the hardware / driver setup and the `WLR_NO_HARDWARE_CURSORS` env var. So maybe the cursor needs to be hidden by SDL but other than that a window that is in fullscreen should cause a direct-scanout to happen.

There is some wlr_scene sample events that could maybe be used to detect if direct scanout is enabled. Starting labwc with the `-d` flag (and redirecting to a file) might also reveal some wlroots internal direct scanout changes.

**Reply by vanfanel** [reply]
This is most interesting, @Consolatis, and I really thank you for your explanations.

Look, this is what I see when RetroArch starts on labwc:

```
00:00:21.087 [types/wlr_compositor.c:692] New wlr_surface 0x55d155b25000 (res 0x55d155b253b0)
00:00:21.087 [types/xdg_shell/wlr_xdg_surface.c:388] new xdg_surface 0x55d155b1a200 (res 0x55d155b1a320)
00:00:21.087 [types/wlr_xdg_decoration_v1.c:219] new xdg_toplevel_decoration 0x55d155b23750 (res 0x55d1559f7ec0)
00:00:21.089 [types/wlr_compositor.c:692] New wlr_surface 0x55d155b270a0 (res 0x55d155b27450)
00:00:21.198 [../src/xdg.c:140] client (org.libretro.RetroArch) did not respond to configure request in 100 ms
00:00:21.638 [types/wlr_pointer_constraints_v1.c:242] new locked_pointer 0x55d155970cf0 (res 0x55d155b35160)
00:00:21.641 [types/wlr_pointer_constraints_v1.c:342] constrained 0x55d155970cf0
00:00:21.641 [../src/view-impl-common.c:50] [map] identifier=org.libretro.RetroArch, title=RetroArch
                                                                                                                                                 
00:00:21.641 [types/scene/wlr_scene.c:1709] Direct scan-out enabled
```

So, yes, Direct scan-out is being used. Fantastic stuff.

This is when RetroArch quits:
```
00:03:11.914 [types/wlr_pointer_constraints_v1.c:352] unconstrained 0x55d155970cf0
00:03:11.914 [types/wlr_pointer_constraints_v1.c:47] destroying constraint 0x55d155970cf0
00:03:11.914 [types/scene/wlr_scene.c:1709] Direct scan-out disabled
```

This is SDLPop starting on labwc, which I use as an SDL2 test:

```
00:04:12.043 [types/wlr_compositor.c:692] New wlr_surface 0x55d155b21d80 (res 0x55d1559ed550)
00:04:14.880 [types/wlr_compositor.c:692] New wlr_surface 0x55d1559e8630 (res 0x55d1559ed5e0)
00:04:14.880 [types/xdg_shell/wlr_xdg_surface.c:388] new xdg_surface 0x55d155b27450 (res 0x55d155b27570)
00:04:14.884 [types/wlr_xdg_decoration_v1.c:219] new xdg_toplevel_decoration 0x55d155b25620 (res 0x55d155b25700)
00:04:14.900 [types/wlr_compositor.c:692] New wlr_surface 0x55d155b289c0 (res 0x55d1559ed5e0)
00:04:14.900 [types/xdg_shell/wlr_xdg_surface.c:388] new xdg_surface 0x55d155b1a200 (res 0x55d155b1a320)
00:04:14.900 [types/wlr_xdg_decoration_v1.c:219] new xdg_toplevel_decoration 0x55d155b253b0 (res 0x55d155b25490)
00:04:14.937 [../src/view-impl-common.c:50] [map] identifier=prince.exe, title=Prince of Persia (SDLPoP) v1.23
                                                                                                                                                 
00:04:14.937 [types/scene/wlr_scene.c:1709] Direct scan-out enabled
```

And this is Scummvm on labwc:
```

00:06:59.233 [types/wlr_compositor.c:692] New wlr_surface 0x55d155b670a0 (res 0x55d155b64420)
00:06:59.269 [types/wlr_compositor.c:692] New wlr_surface 0x55d155b23290 (res 0x55d155b23640)
00:06:59.293 [types/wlr_compositor.c:692] New wlr_surface 0x55d155b23290 (res 0x55d155b23640)
00:06:59.334 [types/wlr_compositor.c:692] New wlr_surface 0x55d155b23290 (res 0x55d155b23640)
00:06:59.358 [types/wlr_compositor.c:692] New wlr_surface 0x55d155b23290 (res 0x55d155b23640)
00:06:59.381 [types/wlr_compositor.c:692] New wlr_surface 0x55d155b23290 (res 0x55d155b23640)
00:07:02.284 [types/wlr_compositor.c:692] New wlr_surface 0x55d155b23290 (res 0x55d155b23640)
00:07:02.284 [types/xdg_shell/wlr_xdg_surface.c:388] new xdg_surface 0x55d1559ed7a0 (res 0x55d1559ed8c0)
00:07:02.284 [types/wlr_xdg_decoration_v1.c:219] new xdg_toplevel_decoration 0x55d155b363c0 (res 0x55d155b253e0)
00:07:02.284 [types/wlr_pointer_constraints_v1.c:242] new confined_pointer 0x55d155b64090 (res 0x55d155b64230)
00:07:02.284 [types/wlr_pointer_constraints_v1.c:47] destroying constraint 0x55d155b64090
00:07:02.284 [types/wlr_pointer_constraints_v1.c:242] new confined_pointer 0x55d155b64090 (res 0x55d155b64230)
00:07:02.284 [types/wlr_pointer_constraints_v1.c:47] destroying constraint 0x55d155b64090
00:07:02.298 [types/wlr_compositor.c:692] New wlr_surface 0x55d155b23290 (res 0x55d155b1b5d0)
00:07:02.298 [types/xdg_shell/wlr_xdg_surface.c:388] new xdg_surface 0x55d155971010 (res 0x55d155b66dd0)
00:07:02.298 [types/wlr_xdg_decoration_v1.c:219] new xdg_toplevel_decoration 0x55d1559f18a0 (res 0x55d155b77090)
00:07:02.311 [types/wlr_pointer_constraints_v1.c:242] new confined_pointer 0x55d155b23060 (res 0x55d155b23640)
00:07:02.398 [../src/xdg.c:140] client (scummvm.exe) did not respond to configure request in 100 ms
00:07:02.408 [types/wlr_pointer_constraints_v1.c:342] constrained 0x55d155b23060
00:07:02.408 [../src/view-impl-common.c:50] [map] identifier=scummvm.exe, title=ScummVM 2.8.0 (Feb  8 2024 00:55:18)
                                                                                                                                                 
00:07:02.408 [types/scene/wlr_scene.c:1709] Direct scan-out enabled
00:07:02.422 [types/wlr_pointer_constraints_v1.c:47] destroying constraint 0x55d155b23060
00:07:02.422 [types/wlr_pointer_constraints_v1.c:242] new confined_pointer 0x55d155b23060 (res 0x55d155b36450)
00:07:02.422 [types/wlr_pointer_constraints_v1.c:342] constrained 0x55d155b23060
00:07:02.557 [types/wlr_pointer_constraints_v1.c:47] destroying constraint 0x55d155b23060
00:07:02.557 [types/wlr_pointer_constraints_v1.c:242] new confined_pointer 0x55d155b23060 (res 0x55d155b1a970)
00:07:02.557 [types/wlr_pointer_constraints_v1.c:342] constrained 0x55d155b23060
00:07:02.557 [types/wlr_pointer_constraints_v1.c:47] destroying constraint 0x55d155b23060
00:07:02.558 [types/wlr_pointer_constraints_v1.c:242] new confined_pointer 0x55d155b23060 (res 0x55d155b23640)
00:07:02.558 [types/wlr_pointer_constraints_v1.c:342] constrained 0x55d155b23060
```
Beautiful, you can also see pointer constraints being activated after Direct Scanout enablement, as expected! This is very important on Scummvm, as it provides good cursor movement when running 4:3 games on 16:9 screens. Not all compositors provides this adequately. Labwc is just perfect on this regard too.

Now, this is the only strange case I found. This is foot (https://codeberg.org/dnkl/foot), the lightweight terminal emulator, being run in fullscreen mode: it seems to cause some strange buffer errors.
```

00:11:58.803 [../src/input/keyboard.c:163] keycode matched
00:11:58.803 [../src/action.c:643] Handling action 5: Execute
warn: main.c:437: 'C' is not a UTF-8 locale, using 'C.UTF-8' instead
00:11:58.813 [types/wlr_compositor.c:692] New wlr_surface 0x55d155b23290 (res 0x55d1559f8340)
00:11:58.813 [types/wlr_compositor.c:692] New wlr_surface 0x55d1559d4e00 (res 0x55d1559ea560)
00:11:58.813 [types/xdg_shell/wlr_xdg_surface.c:388] new xdg_surface 0x55d1559f7cc0 (res 0x55d1559f5a10)
00:11:58.813 [types/wlr_compositor.c:692] New wlr_surface 0x55d1559f64a0 (res 0x55d155b23130)
00:11:58.829 [../src/view-impl-common.c:50] [map] identifier=foot, title=Command Console
                                                                                                                                                 
00:11:58.829 [backend/drm/renderer.c:327] Failed to get DMA-BUF from buffer
00:11:58.829 [backend/drm/drm.c:582] connector HDMI-A-1: Failed to import buffer for scan-out
00:11:58.831 [../src/input/cursor.c:189] set xcursor to shape text
00:11:58.836 [backend/drm/renderer.c:327] Failed to get DMA-BUF from buffer
00:11:58.836 [backend/drm/drm.c:582] connector HDMI-A-1: Failed to import buffer for scan-out
00:11:58.853 [backend/drm/renderer.c:327] Failed to get DMA-BUF from buffer
00:11:58.853 [backend/drm/drm.c:582] connector HDMI-A-1: Failed to import buffer for scan-out
00:11:59.331 [backend/drm/renderer.c:327] Failed to get DMA-BUF from buffer
00:11:59.331 [backend/drm/drm.c:582] connector HDMI-A-1: Failed to import buffer for scan-out
00:11:59.831 [backend/drm/renderer.c:327] Failed to get DMA-BUF from buffer
00:11:59.831 [backend/drm/drm.c:582] connector HDMI-A-1: Failed to import buffer for scan-out
```

The two "Failed" lines repeat infinitely while foot is running.
What do you think? Should I report this to the foot author, or is this something expected somehow?

**Reply by Consolatis** [reply]
Hm.. interesting. Maybe foot uses some unsupported DRM modifiers for its buffer.
Could you give this PR a try to see if it changes anything?
- #1278

**Reply by vanfanel** [reply]
Of course. I tried that PR and there's no difference with Foot:

```
00:00:23.103 [../src/input/keyboard.c:174] keycode matched
00:00:23.103 [../src/action.c:674] Handling action 5: Execute
warn: main.c:437: 'C' is not a UTF-8 locale, using 'C.UTF-8' instead
00:00:23.113 [types/wlr_compositor.c:692] New wlr_surface 0x55667c732060 (res 0x55667c732410)
00:00:23.113 [types/wlr_compositor.c:692] New wlr_surface 0x55667c72bf30 (res 0x55667c72c2e0)
00:00:23.113 [types/xdg_shell/wlr_xdg_surface.c:388] new xdg_surface 0x55667c6a5400 (res 0x55667c72a3b0)
00:00:23.113 [types/wlr_compositor.c:692] New wlr_surface 0x55667c7301a0 (res 0x55667c713c10)
00:00:23.127 [../src/view-impl-common.c:49] [map] identifier=foot, title=Command Console
                                                                                                                                                 
00:00:23.127 [backend/drm/renderer.c:327] Failed to get DMA-BUF from buffer
00:00:23.127 [backend/drm/drm.c:582] connector HDMI-A-1: Failed to import buffer for scan-out
00:00:23.129 [../src/input/cursor.c:189] set xcursor to shape text
00:00:23.140 [backend/drm/renderer.c:327] Failed to get DMA-BUF from buffer
00:00:23.140 [backend/drm/drm.c:582] connector HDMI-A-1: Failed to import buffer for scan-out
00:00:23.157 [backend/drm/renderer.c:327] Failed to get DMA-BUF from buffer
00:00:23.157 [backend/drm/drm.c:582] connector HDMI-A-1: Failed to import buffer for scan-out
00:00:23.629 [backend/drm/renderer.c:327] Failed to get DMA-BUF from buffer
00:00:23.629 [backend/drm/drm.c:582] connector HDMI-A-1: Failed to import buffer for scan-out
00:00:24.129 [backend/drm/renderer.c:327] Failed to get DMA-BUF from buffer
00:00:24.129 [backend/drm/drm.c:582] connector HDMI-A-1: Failed to import buffer for scan-out
00:00:24.629 [backend/drm/renderer.c:327] Failed to get DMA-BUF from buffer
00:00:24.629 [backend/drm/drm.c:582] connector HDMI-A-1: Failed to import buffer for scan-out
00:00:25.129 [backend/drm/renderer.c:327] Failed to get DMA-BUF from buffer
```

..etc

**Reply by Consolatis** [reply]
Thanks for testing. I am not sure what is going on there with foot.
We could absolutely miss something on our end (although based on the logs its all wlroots internal) but not sure what it could be.
Opening an issue over at foot and asking for a potential explanation might indeed help (and maybe include on what hardware this occurs if its something like a raspi).

Does it only occur when starting foot in fullscreen?

**Reply by vanfanel** [reply]
I will open an issue there, yes.

This happens on a common x86_64 computer, based on the Intel N100 (which has Intel XE integrated graphics, pretty common stuff I guess).

Also, it **only** happens when Foot runs fullscreen or maximized (it's configuration file allows both values).

**Reply by vanfanel** [reply]
Issue opened on Foot repository: https://codeberg.org/dnkl/foot/issues/1605

**Reply by vanfanel** [reply]
@Consolatis seems that Foot is a pure SHM program, could that be preventing Direct-scanout from working?

**Reply by Consolatis** [reply]
Yes, that seems to be the case. I wasn't aware that foot is only supporting SHM buffers (which obviously is totally fine for a terminal emulator).

**Reply by vanfanel** [reply]
Then that's about It. Case closed! Thanks for this rich discussion that tough me many things!
